velox/functions/sparksql/String.h

```diff
@@ -1181,4 +1181,63 @@ struct FindInSetFunction {
   }
 };
 
+template <typename T>
+struct SoundexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  static constexpr bool is_default_ascii_behavior = true;
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  void call(out_type<Varchar>& result, const arg_type<Varchar>& input) {
+    auto inputData = reinterpret_cast<const uint8_t*>(input.data());
+    size_t inputSize = input.size();
+    if (inputSize == 0) {
+      result.setEmpty();
+      return;
+    }
+    auto b = inputData[0];
+    if ('a' <= b && b <= 'z') {
+      b -= 32;
+    } else if (b < 'A' || 'Z' < b) {
+      // First character must be a letter, otherwise input is returned.
```

Comment:
Maybe document this limitation. Can we use `std::isalpha` to check the character?

Reply:
I think this is the requirement from the algorithm.